File: leetcode-js/056-merge-intervals.js
/**
 * Definition for an interval.
 * function Interval(start, end) {
 *     this.start = start;
 *     this.end = end;
 * }
 */
/**
 * @param {Interval[]} intervals
 * @return {Interval[]}
 */
var merge = function(intervals) {
    if (intervals.length <= 1) {
        return intervals;
    }

    intervals.sort((a, b) => a.start - b.start);

    const result = [];
    let {start, end} = intervals[0];

    for (const interval of intervals.slice(1)) {
        if (interval.start <= end) {
            // When next interval starts before it ends, extend the ending
            end = Math.max(end, interval.end);
        } else {
            // When no overlap, merge previous overlapped intervals
            result.push(new Interval(start, end));

            // Set new start and end
            start = interval.start;
            end = interval.end;
        }
    }

    result.push(new Interval(start, end));

    return result;
};

File: leetcode-java/056-merge-intervals.java
/**
 * Definition for an interval.
 * public class Interval {
 *     int start;
 *     int end;
 *     Interval() { start = 0; end = 0; }
 *     Interval(int s, int e) { start = s; end = e; }
 * }
 */
public class Solution {
    public List<Interval> merge(List<Interval> intervals) {
        if (intervals.size() == 0) {
            return new LinkedList<>();
        }

        Collections.sort(intervals, new Comparator<Interval>() {
            @Override
            public int compare(Interval i1, Interval i2) {
                return Integer.compare(i1.start, i2.start);
            }
        });

        List<Interval> res = new LinkedList<>();
        Interval curr = new Interval(intervals.get(0).start, intervals.get(0).end);

        for (int i = 1; i < intervals.size(); i++) {
            if (curr.end < intervals.get(i).start) {
                res.add(curr);
                curr = new Interval(intervals.get(i).start, intervals.get(i).end);
            } else {
                curr.end = Math.max(curr.end, intervals.get(i).end);
            }
        }

        res.add(curr);

        return res;
    }
}
